guest_libs/crypto/src/secp256k1.rs

```diff
@@ -1,11 +1,172 @@
+        bigint::CheckedAdd,
+        ops::{Invert, Reduce},
+        sec1::{FromEncodedPoint, ToEncodedPoint},
+    },
+};
+
+type UntaggedUncompressedPoint = [u8; 64];
+
+#[repr(align(4))]
+struct Aligned64([u8; 64]);
 
 /// secp256k1 ECDSA signature recovery.
 #[inline]
 pub fn secp256k1_ecrecover(
-    _sig: &[u8; 64],
-    _recid: u8,
-    _msg: &[u8; 32],
+    sig: &[u8; 64],
+    recid: u8,
+    msg: &[u8; 32],
 ) -> Result<[u8; 32], CenoCryptoError> {
-    unimplemented!()
+    // Copied from <https://github.com/alloy-rs/alloy/blob/8e9be40eb0e7c27618db1316989f77f1cfe3accb/crates/consensus/src/crypto.rs#L311-L334>
+    let mut signature = Signature::from_slice(sig)?;
+    let mut recid = recid;
+
+    // normalize signature and flip recovery id if needed.
+    if let Some(sig_normalized) = signature.normalize_s() {
+        signature = sig_normalized;
+        recid ^= 1;
+    }
+    let recid = RecoveryId::from_byte(recid).expect("recovery ID is valid");
+
+    // recover key
+    let recovered_key = recover_from_prehash_unchecked(&msg[..], &signature, recid)?;
+
+    let mut hasher = Keccak::v256();
+    hasher.update(&recovered_key);
+    let mut hash = [0u8; 32];
+    hasher.finalize(&mut hash);
+    // truncate to 20 bytes
+    hash[..12].fill(0);
+    Ok(hash)
+}
+
+/// Copied from <https://github.com/RustCrypto/signatures/blob/89232d6a962a199fd8211a117db74408353e4383/ecdsa/src/recovery.rs#L278-L316>
+/// Modified to use ceno syscalls
+fn recover_from_prehash_unchecked(
+    prehash: &[u8],
+    signature: &Signature,
+    recovery_id: RecoveryId,
+) -> k256::ecdsa::signature::Result<UntaggedUncompressedPoint> {
+    let (r, s) = signature.split_scalars();
+    let z = <Scalar as Reduce<U256>>::reduce_bytes(&bits2field::<Secp256k1>(prehash)?);
+
+    let mut r_bytes = r.to_repr();
+    if recovery_id.is_x_reduced() {
+        let decoded: U256 = FieldBytesEncoding::<Secp256k1>::decode_field_bytes(&r_bytes);
+        match decoded.checked_add(&Secp256k1::ORDER).into_option() {
+            Some(restored) => {
+                r_bytes = <U256 as FieldBytesEncoding<Secp256k1>>::encode_field_bytes(&restored)
+            }
+            // No reduction should happen here if r was reduced
+            None => return Err(Error::new()),
+        };
+    }
+
+    // Modified part: use ceno syscall to decompress point
+    // Original:
+    // let R = AffinePoint::decompress(&r_bytes, u8::from(recovery_id.is_y_odd()).into());
+    let r_point = {
+        let mut buf = Aligned64([0u8; 64]);
+        buf.0[..32].copy_from_slice(&r_bytes);
+
+        // SAFETY:
+        // [x] The input array should be 64 bytes long, with the first 32 bytes containing the X coordinate in
+        //   big-endian format.
+        // [x] The caller must ensure that `point` is valid pointer to data that is aligned along a four byte
+        // boundary.
+        ceno_rt::syscalls::syscall_secp256k1_decompress(&mut buf.0, recovery_id.is_y_odd());
```

Comment:
Hi, we do not have decompress function yet, does it work if we make it native guest program? such as comment above

```
let R = AffinePoint::decompress(&r_bytes, u8::from(recovery_id.is_y_odd()).into());
```